### PR TITLE
limiter: support workloadselector

### PR DESCRIPTION
**Pull request by MouceL**
support workloadSelector , the priority is  workloadSelector > name.namespace

the limiter rule will dispatch to pods which has labels  `app: productpage, env: test ` in same ns with SmartLimiter.

if SmartLimiter apply to istio root ns, it will match all pods in cluster with  `app: productpage, env: test `

```yaml
apiVersion: microservice.slime.io/v1alpha2
kind: SmartLimiter
metadata:
  name: productpage
  namespace: default
spec:
  workloadSelector:
    app: productpage
    env: test  
  sets:
    _base:
      descriptor:
      - action:
          fill_interval:
            seconds: 10
          quota: "2"
          strategy: "single"
        condition: "true"
        target:
          port: 9080
```

### CODE REVIEW

File: staging/src/slime.io/slime/modules/limiter/controllers/smartlimiter_controller.go [hunk 1]
```diff
@@ -197,15 +196,33 @@ func NewProducerConfig(env bootstrap.Environment, cfg *config.Limiter) (*metric.
...
 
-	// query hostname base on instance
-	host := buildHost(instance, out)
-
 	// set interest mapping
-	r.interest.Set(key, host)
-	log.Infof("set %s/%s = %s", instance.Namespace, instance.Name, host)
+	meta := SmartLimiterMeta{}
+
+	if outbound {
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/limiter/controllers/smartlimiter_controller.go:205]: not easy to understand

File: staging/src/slime.io/slime/modules/limiter/controllers/producer.go [hunk 2]
```diff
@@ -161,14 +192,18 @@ func (r *SmartLimiterReconciler) handleLocalEvent(host string, loc types.Namespa
...
 	if r.env.Config == nil || r.env.Config.Metric == nil || r.env.Config.Metric.Prometheus == nil || r.env.Config.Metric.Prometheus.Handlers == nil {
 		log.Infof("query handler is empty, skip query")
 		return nil
 	}
 	handlers := r.env.Config.Metric.Prometheus.Handlers
 
-	if !strings.HasSuffix(host, util.WellknownK8sSuffix) {
+	// TODO workloadSelector
+	if len(LimiterMeta.workloadSelector) > 0 {
+		log.Info("promql is closed when workloadSelector is specified")
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/limiter/controllers/producer.go:204]: warn?

Comment (YonkaFang) [reply staging/src/slime.io/slime/modules/limiter/controllers/producer.go:204]: log with namespacedname?

File: staging/src/slime.io/slime/modules/limiter/controllers/producer.go [hunk 3]
```diff
@@ -72,45 +71,76 @@ func (r *SmartLimiterReconciler) handleTickerEvent(event trigger.TickerEvent) me
...
+		subsetInfo[util.WellknownBaseSet] = len(pods)
+
+		meta := generateMeta(subsetInfo, loc)
+		metaInfo := meta.String()
+		log.Debugf("get workloadSelector meta info %s", metaInfo)
+		if metaInfo == "" {
+			return nil
+		}
+		queryMap[metaInfo] = []metric.Handler{}
+		return queryMap
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/limiter/controllers/producer.go:132]: we can handle different cases in separated functions

File: staging/src/slime.io/slime/modules/limiter/controllers/envoy_filters.go [hunk 4]
```diff
@@ -60,12 +59,23 @@ func (r *SmartLimiterReconciler) GenerateEnvoyConfigs(spec microservicev1alpha2.
...
+		return setsEnvoyFilter, setsSmartLimitDescriptor, globalDescriptors, nil
+	}
+
+	meta, ok := v.(SmartLimiterMeta)
+	if !ok {
+		log.Error("covert meta to SmartLimiterMeta err")
+		return setsEnvoyFilter, setsSmartLimitDescriptor, globalDescriptors, nil
+	}
+
+	// subset feature is only used when host/seHost is specified
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/limiter/controllers/envoy_filters.go:73]: comment can by more comprehensive and with more business meanings

File: staging/src/slime.io/slime/modules/limiter/controllers/smartlimiter_controller.go [hunk 5]
```diff
@@ -197,15 +196,33 @@ func NewProducerConfig(env bootstrap.Environment, cfg *config.Limiter) (*metric.
...
+		meta.outbound = true
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/limiter/controllers/smartlimiter_controller.go:206]: why outbound do not set workloadselector?

File: staging/src/slime.io/slime/modules/limiter/controllers/smartlimiter_controller.go (hunk 1 above)
Comment (MouceL) [reply staging/src/slime.io/slime/modules/limiter/controllers/smartlimiter_controller.go:205]: Simplified

File: staging/src/slime.io/slime/modules/limiter/controllers/producer.go (hunk 3 above)
Comment (MouceL) [reply staging/src/slime.io/slime/modules/limiter/controllers/producer.go:132]: done

**Pull request merged**